### PR TITLE
GUVNOR-2826: Prevent same binding for fact and field

### PR DESCRIPTION
**Pull request by jomarko**

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopup.java [hunk 1]
```diff
@@ -162,32 +162,25 @@ public void applyChanges() {
...
 
         //Check for unique binding
-        if ( isNew ) {
-            if ( editingCol.isBound() && !isBindingUnique( editingCol.getBinding() ) ) {
-                Window.alert( GuidedDecisionTableConstants.INSTANCE.PleaseEnterANameThatIsNotAlreadyUsedByAnotherPattern() );
+        String factBinding = editingPattern.getBoundName();
+        String fieldBinding = editingCol.getBinding();
+        if( factBinding != null && fieldBinding != null ) {
+            if ( editingCol.isBound() && ( !isBindingUnique( fieldBinding ) || factBinding.compareTo( fieldBinding ) == 0 ) ) {
+                view.warnAboutAlreadyUsedBinding();
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopup.java]: This change could be covered by tests.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 2]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
         verify( fieldChangePopUp ).hide();
         verify( popup.view ).enableFooter( true );
     }
+
+    @Test
+    public void testApplyChangesMissingHeader() {
+        column.setHeader( null );
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 3]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
...
+
+        popup.applyChanges();
+        verify( popup.view ).warnAboutMissingColumnHeaderDescription();
+    }
+
+    @Test
+    public void testApplyChangesMissingFactField() {
+        column.setFactField( null );
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 4]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
...
+
+        popup.applyChanges();
+        verify( popup.view ).warnAboutMissingFactField();
+    }
+
+    @Test
+    public void testApplyChangesMissingOperator() {
+        column.setOperator( null );
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 5]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
...
+
+        popup.applyChanges();
+        verify( popup.view ).warnAboutMissingOperator();
+    }
+
+    @Test
+    public void testApplyChangesPredicate() {
+        column.setConstraintValueType( BaseSingleFieldConstraint.TYPE_PREDICATE );
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 6]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
...
+        verify( popup.view ).warnAboutAlreadyUsedColumnHeaderName();
+    }
+
+    @Test
+    public void testClearBindingWhenPredicate() throws Exception {
+        column.setBinding( "$field" );
+        column.setConstraintValueType( BaseSingleFieldConstraint.TYPE_PREDICATE );
+
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java [hunk 7]
```diff
@@ -278,4 +280,107 @@ public void testCancelFieldChangePopUp() throws Exception {
...
+
+        BaseColumn baseColumn = mock( BaseColumn.class );
+        when( baseColumn.getHeader() ).thenReturn( COLUMN_HEADER );
+        CompositeColumn compositeColumn = mock( CompositeColumn.class );
+        when( compositeColumn.getChildColumns() ).thenReturn(Arrays.asList( baseColumn ) );
+
+        model.getConditions().add( compositeColumn );
+
+        setUpPopup( model,
+                pattern,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopupTest.java]: Formatting problem.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopup.java (hunk 1 above)
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/ConditionPopup.java]: There are now 4 test methods `testBinding*` that should cover all cases. Let me know what do you think.

**Comment by jomarko** [comment]
@karreiro Hi, please could you have a look if the coverage is enough now?

**Comment by karreiro** [comment]
Hi @jomarko. The coverage looks good :-)

Thanks!

**Pull request merged**